design-proposals/sriov-interfaces-mapping.md

```diff
@@ -0,0 +1,694 @@
+| VM creation/restart/migration<br/>(new virt-launcher) | 3. Regression<br/>Virt-handler sends the network-to-pci mapping information to virt-launcher through gRPC, virt-launcher should fall back to legacy mapping method. | 4. No regression                                                                                                                                                                                                                                                        |
+
+#### Conclusion
+- (2) and (3):<br/>
+There might be a disruption to VMs when they were created:
+  - After the virt-controller is updated (new virt-launcher), on a node with an old instance of virt-handler.
+  - Before the virt-controller is updated (old virt-launcher), on a node with a new instance of virt-handler.
+  <br/><br/>
+  Since the gRPC command server (virt-launcher) and client (virt-handler) runs with different API versions, virt-launcher may fail to synchronize with the new VMI state.
+  <br/><br/>
+
+- (4) When both virt-handler and virt-controller are updated, new VMs will be created with the correct SR-IOV interfaces assignment.
+
+### Pros
+1. No need to involve virt-controller.
+2. No need to create annotations on the pod or VMI. 
+
+### Cons
+1. Requires API changes to virt-handler to virt-launcher commands API [[7]](https://github.com/kubevirt/kubevirt/blob/1a8f08103e48c5f2bb2f5826d118507ce7ec1f0c/pkg/handler-launcher-com/cmd/v1/cmd.pb.go) [[8]](https://github.com/kubevirt/kubevirt/blob/1a8f08103e48c5f2bb2f5826d118507ce7ec1f0c/pkg/handler-launcher-com/cmd/v1/cmd.pb.go#L334).
+2. virt-handler needs to fetch the pod object in order to get the mapping information.
+
+## 4. Expose network-to-pci mapping through the VMI status
+This option uses the same basic ideas from [option 2](#2.virt-controller-exposes-network-to-pci-mapping-through-two-VMI-annotations) but instead of using VMI annotations it
+uses VMI Status additional fields to persist the mapping.
+
+1. Extend the VMI status API to reflect each SR-IOV interface underlying PCI address.
+
+2. The virt-controller updates the status fields with the mapping information.
+
+3. virt-launcher VMI extract the network-to-pci mapping from the VMI status and create each HostDevice accordingly.<br/>
+   In case there is not enough information in the VMI status (i.e: missing host PCI address) fall back to legacy mapping method. 
+
+#### Pros
+1. No need for annotations.
+
+#### Cons
+1. The status field will have to support mapping from two pods, supporting migration.
+2. It is confusing to use the VMI status as an input to the virt-launcher.
+3. The host PCI address is now exposed to the user and may be confusing.
```

Comment:
I wonder if it is more confusing than huge annotations on pods. We already track a lot of internal stuff there which users can not directly use for anything specific, like migration ports and so forth. I would not worry about that.

Reply:
As this option suggests to add this information to the vmi-status, I think it would be rather confusing that the user configures the vmi with guest PCI-Address, like so:
```
kind: VirtualMachine
…
spec:
  domain:
    devices:
      interfaces:
      - name: sriovnet1
        pciAddress: 0000:00:01.0
        sriov: {}
    …
  networks:
  - name: sriovnet1
    multus:
      networkName: default/sriov-network-vlan100
```
(`0000:00:01.0` being the pci-address as it will appear on the guest)

and the vmi-status would show something like this 
```
apiVersion: v1
items:
- apiVersion: kubevirt.io/v1
  kind: VirtualMachineInstance
...
  spec:
...
  status:
    guestOSInfo:
...
    interfaces:
    - infoSource: domain, guest-agent
      interfaceName: eth0
...
    - infoSource: domain, guest-agent
      interfaceName: eth1
      ipAddress: fe80::a8bb:ccff:fedd:ee01
      ipAddresses:
      - fe80::a8bb:ccff:fedd:ee01
      hostPciAddress: 0000:40:20.0
```

My point being, this pci-address is for internal use, and thus should not appear in the vmi-status. 
Also consider the complication of how to expose this while in migration.

Reply:
> My point being, this pci-address is for internal use, and thus should not appear in the vmi-status.

Yes I understand your concern. I would just not mix it. I would favor something like this:

```
status:
  nodeInfo:
     "node01":
        interfaces:
          [...]
     "node02":
        [...]
```

and just clear the whole entry from there after a migration success or failure. The annotation proposal may be even simpler, but this is simple too and it follows all conventions with the advantages mentioned.

As said, we already include a lot of info regarding to migrations and so forth which are not relevant for users either.

Reply:
Especially, if this route is given, we will never have to differentiate or change the info flow on this level not matter what comes in the future. The other approach sounds tempting now, but does not guarantee at all that this approach is not needed too at some point.

As I understand it, the proposals were mostly driven by the assumption that metadata has to be used one or the other way, since info like this can't be in the vmi status or the pod status. At least that is how I understood the feedback on the PR. I would reconsider it with the new info provided.